pymeasure/instruments/toptica/ibeamsmart.py

```diff
@@ -22,99 +22,267 @@
+import logging
+import re
+import time
+from warnings import warn
+
+from pyvisa.errors import VisaIOError
+
+from pymeasure.instruments import Channel, Instrument
 from pymeasure.instruments.validators import strict_discrete_set, strict_range
 
 
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+def _deprecation_warning(property_name):
+    def func(x):
+        warn(f'Deprecated property name "{property_name}", use the channels '
+             '"enabled" property instead.', FutureWarning)
+        return x
+    return func
+
+
+class DriverChannel(Channel):
+    """A laser diode driver channel for the IBeam Smart laser."""
+
+    power = Channel.setting(
+        "ch {ch} pow %f mic",
+        """Set the output power in µW (float up to 200000).""",
+        check_set_errors=True,
+        validator=strict_range,
+        values=[0, 200000],
+    )
+
+    enabled = Channel.control(
+        "sta ch {ch}",
+        "%s {ch}",
+        """Control the enabled state of the driver channel.""",
+        validator=strict_discrete_set,
+        values=[True, False],
+        get_process=lambda s: True if s == 'ON' else False,
+        set_process=lambda v: "en" if v else "di",
+        check_set_errors=True,
+    )
+
+
 class IBeamSmart(Instrument):
     """ IBeam Smart laser diode
 
-    :param port: pyvisa resource name of the instrument
-    :param baud_rate: communication speed, defaults to 115200
-    :param kwargs: Any valid key-word argument for VISAAdapter
+    For the usage of the different diode driver channels, see the manual
+
+    .. code::
+
+        laser = IBeamSmart("SomeResourceString")
+        laser.emission = True
+        laser.ch_2.power = 1000  # µW
+        laser.ch_2.enabled = True
+        laser.shutdown()
+
+    :param adapter: pyvisa resource name or adapter instance.
+    :param \\**kwargs: Any valid key-word argument for VISAAdapter.
     """
-    def __init__(self, adapter, baud_rate=115200, **kwargs):
-        if not isinstance(adapter, str):
-            raise TypeError("'adapter' must be a pyvisa resource name string")
+    _reg_value = re.compile(r"\w+\s+=\s+(\w+)")
+
+    channels = Instrument.ChannelCreator(DriverChannel, (1, 2, 3, 4, 5))
+
+    def __init__(self, adapter, name="Toptica IBeam Smart laser diode", **kwargs):
         super().__init__(
-            TopticaAdapter(adapter, baud_rate, **kwargs),
-            "toptica IBeam Smart laser diode",
+            adapter,
+            name,
             includeSCPI=False,
+            read_termination='\r\n',
+            write_termination='\r\n',
+            baud_rate=115200,
+            **kwargs
         )
+        # configure communication mode: no repeating and no command prompt
+        self.write('echo off')
+        self.write('prom off')
+        time.sleep(0.04)
+        # clear the initial messages from the controller
+        try:
+            self.adapter.flush_read_buffer()
+        except AttributeError:
```

Comment:
Here I see a problem: I guess with `ProtocolAdapter` you get an `AttributeError`. But if I use a translator device (TCPIP to RS232 via VisaAdapter) I get an `NotImplementedError`. So I guess one either should make `ProtocolAdapter` also raise the same exception or catch here both types of Exceptions.

```suggestion
        except (AttributeError, NotImplementedError):
```

In case a translator device is used, one would need here also a fallback implementation of the flush method.
e.g. something like this. But I would find this more elegant to be handled on the Adapter level...
```
def _fallback_flush_read_buffer(self):
    timeout = self.adapter.connection.timeout
    self.adapter.connection.timeout = 1
    try:
        while True:
            super().read()
    except VisaIOError:
        pass
    self.adapter.connction.timeout = timeout
```

Reply:
> But if I use a translator device (TCPIP to RS232 via VisaAdapter) I get an ` NotImplementedError`

On what, exactly? Can you paste the error? `VisaAdapter` has `flush_read_buffer()`, so I'd be curious what it's tripping ober?

Reply:
> In case a translator device is used, one would need here also a fallback implementation of the flush method.
e.g. something like this. But I would find this more elegant to be handled on the Adapter level...

Could you elaborate here, too? In my copy of pyvisa, `TCPIPInstrument` should get its `flush` from `MessageBasedResource`, so I'm not sure why we need to have a fallback for that? 
It's a bit hard to debug stuff like this from far away -- exact repros and error messages help :sweat_smile:

Reply:
> > But if I use a translator device (TCPIP to RS232 via VisaAdapter) I get an ` NotImplementedError`
> 
> On what, exactly? Can you paste the error? `VisaAdapter` has `flush_read_buffer()`, so I'd be curious what it's tripping over?

Since this is independent of the device, you can reproduce the error yourself using:
```
import pymeasure.adapters
test = pymeasure.adapters.VISAAdapter("TCPIP::127.0.0.1::8080::SOCKET")
test.flush_read_buffer()
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[4], line 1
----> 1 test.flush_read_buffer()

File ~/tmp/pymeasure/pymeasure/adapters/visa.py:228, in VISAAdapter.flush_read_buffer(self)
    221 def flush_read_buffer(self):
    222     """ Flush and discard the input buffer
    223 
    224     As detailed by pyvisa, discard the read buffer contents and if data was present
    225     in the read buffer and no END-indicator was present, read from the device until
    226     encountering an END indicator (which causes loss of data).
    227     """
--> 228     self.connection.flush(pyvisa.constants.BufferOperation.discard_read_buffer)

File ~/.local/lib/python3.10/site-packages/pyvisa/resources/messagebased.py:797, in MessageBasedResource.flush(self, mask)
    784 def flush(self, mask: constants.BufferOperation) -> None:
    785     """Manually clears the specified buffers.
    786 
    787     Depending on the value of the mask this can cause the buffer data
   (...)
    795 
    796     """
--> 797     self.visalib.flush(self.session, mask)

File ~/.local/lib/python3.10/site-packages/pyvisa_py/highlevel.py:222, in PyVisaLibrary.flush(self, session, mask)
    220 except KeyError:
    221     return self.handle_return_value(session, StatusCode.error_invalid_object)
--> 222 return self.handle_return_value(session, sess.flush(mask))

File ~/.local/lib/python3.10/site-packages/pyvisa_py/sessions.py:439, in Session.flush(self, mask)
    418 def flush(self, mask: constants.BufferOperation) -> StatusCode:
    419     """Flush the specified buffers.
    420 
    421     The buffers can be associated with formatted I/O operations and/or
   (...)
    437 
    438     """
--> 439     raise NotImplementedError

NotImplementedError: 
```

So somehow (at least when using `pyvisa-py`) the flush method seems not to work as you expect. This is in my opinion also the reason why the fallback would be beneficial.

Reply:
It seems pyvisa-py just implemented flush for the serial device. https://github.com/pyvisa/pyvisa-py/pull/208

Reply:
Maybe improving pyvisa-py might be better than creating it ourselves?

Reply:
I agree that having this in pyvisa-py would be the best, but I am not sure if this is gonna happen quickly for many interface types. I can open an issue there, but it seems to me best would be to have a fallback implementation in pymeasure. It should be used only in case of a `NotImplementedError` (or alike) of the underlying code.

Reply:
We could (and should) add the error handling in the `VISAAdapter.flush_read_buffer` method, not here, until pyvisa-py is updated. Please feel free to open an issue (or directly a PR).

The instrument will work flawlessly without this fix. I'll resolve.